File: backend/app/parsers/generic_csv.py
"""Generic CSV template import for unsupported brokers/sources.

Documented columns (header row, case-insensitive; extra columns ignored):
    name, type, isin, symbol, scheme_code, quantity, avg_cost,
    invested_value, current_value, amc, plan, folio, sector, market_cap, asset_class

Only ``name`` and one of (``current_value`` or ``quantity``) are strictly required.
``type`` should be one of the InstrumentType values (stock, etf, mutual_fund, sgb, bond,
gsec, fd, ppf, nps, reit, invit, cash, digital_gold, real_estate, other); defaults to
``other``. ``asset_class`` is an optional explicit override applied during classification.
"""

from __future__ import annotations

from app.domain.taxonomy import InstrumentType, Source
from app.parsers.base import ParsedHolding, ParseResult
from app.parsers.csv_utils import find_col, sniff_rows, to_float

_VALID_TYPES = {t.value for t in InstrumentType}


def parse(content: str, file_name: str | None = None, account_name: str = "Imported") -> ParseResult:
    result = ParseResult(source=Source.GENERIC_CSV, file_name=file_name)
    rows = sniff_rows(content)
    if not rows:
        result.error("No data rows found in file.")
        return result

    h = list(rows[0].keys())
    cols = {
        "name": find_col(h, "name", "instrument", "scheme", "description"),
        "type": find_col(h, "type", "instrumenttype", "assettype"),
        "isin": find_col(h, "isin"),
        "symbol": find_col(h, "symbol", "ticker"),
        "scheme_code": find_col(h, "schemecode", "amfi", "amficode"),
        "quantity": find_col(h, "quantity", "qty", "units"),
        "avg_cost": find_col(h, "avgcost", "averageprice", "avgprice"),
        "invested_value": find_col(h, "investedvalue", "invested", "buyvalue", "cost"),
        "current_value": find_col(h, "currentvalue", "value", "marketvalue", "amount"),
        "amc": find_col(h, "amc", "fundhouse"),
        "plan": find_col(h, "plan"),
        "folio": find_col(h, "folio"),
        "sector": find_col(h, "sector"),
        "market_cap": find_col(h, "marketcap", "cap"),
        "asset_class": find_col(h, "assetclass", "class"),
    }
    if not cols["name"]:
        result.error("Generic CSV needs a 'name' column.")
        return result

    for row in rows:
        name = (row.get(cols["name"]) or "").strip()
        if not name:
            continue
        raw_type = (row.get(cols["type"]) or "").strip().lower() if cols["type"] else ""
        itype = raw_type if raw_type in _VALID_TYPES else InstrumentType.OTHER.value

        qty = to_float(row.get(cols["quantity"])) if cols["quantity"] else None
        curval = to_float(row.get(cols["current_value"])) if cols["current_value"] else None
        if curval is None and qty is None:
            result.warn(f"Skipped {name}: no quantity or current value.", context=name)
            continue

        result.holdings.append(
            ParsedHolding(
                name=name,
                instrument_type=InstrumentType(itype),
                isin=(row.get(cols["isin"]) or "").strip().upper() or None if cols["isin"] else None,
                symbol=(row.get(cols["symbol"]) or "").strip().upper() or None if cols["symbol"] else None,
                scheme_code=(row.get(cols["scheme_code"]) or "").strip() or None if cols["scheme_code"] else None,
                quantity=qty or 1.0,
                avg_cost=to_float(row.get(cols["avg_cost"])) if cols["avg_cost"] else None,
                invested_value=to_float(row.get(cols["invested_value"])) if cols["invested_value"] else None,
                current_value=curval,
                amc=(row.get(cols["amc"]) or "").strip() or None if cols["amc"] else None,
                plan=(row.get(cols["plan"]) or "").strip().lower() or None if cols["plan"] else None,
                folio=(row.get(cols["folio"]) or "").strip() or None if cols["folio"] else None,
                source=Source.GENERIC_CSV,
                account_name=account_name,
                account_kind="manual",
                sector_hint=(row.get(cols["sector"]) or "").strip() or None if cols["sector"] else None,
                market_cap_hint=(row.get(cols["market_cap"]) or "").strip() or None if cols["market_cap"] else None,
                category_hint=(row.get(cols["asset_class"]) or "").strip() or None if cols["asset_class"] else None,
                raw=row,
            )
        )

    result.info(f"Parsed {len(result.holdings)} rows from generic CSV.")
    return result

File: frontend/src/engine/parse/genericCsv.ts
// Generic CSV template → TS mirror of backend/app/parsers/generic_csv.py.

import { InstrumentType, Source } from "../taxonomy";
import { ParsedTxn, ParseResult, holding } from "./types";
import { findCol, parseDate, round2, sniffRows, toFloat } from "./csv";

const VALID_TYPES = new Set<string>(Object.values(InstrumentType));

export function parseGenericCsv(content: string, fileName: string | null = null, accountName = "Imported"): ParseResult {
  const result = new ParseResult(Source.GENERIC_CSV, fileName);
  const rows = sniffRows(content);
  if (!rows.length) {
    result.error("No data rows found in file.");
    return result;
  }
  const h = Object.keys(rows[0]);
  const cols = {
    name: findCol(h, "name", "instrument", "scheme", "description"),
    type: findCol(h, "type", "instrumenttype", "assettype"),
    isin: findCol(h, "isin"),
    symbol: findCol(h, "symbol", "ticker"),
    scheme_code: findCol(h, "schemecode", "amfi", "amficode"),
    quantity: findCol(h, "quantity", "qty", "units"),
    avg_cost: findCol(h, "avgcost", "averageprice", "avgprice"),
    invested_value: findCol(h, "investedvalue", "invested", "buyvalue", "cost"),
    current_value: findCol(h, "currentvalue", "value", "marketvalue", "amount"),
    amc: findCol(h, "amc", "fundhouse"),
    plan: findCol(h, "plan"),
    folio: findCol(h, "folio"),
    sector: findCol(h, "sector"),
    market_cap: findCol(h, "marketcap", "cap"),
    asset_class: findCol(h, "assetclass", "class"),
    start_date: findCol(h, "startdate", "purchasedate", "investmentdate", "date"),
  };
  if (!cols.name) {
    result.error("Generic CSV needs a 'name' column.");
    return result;
  }
  const g = (row: Record<string, string>, key: string | null) => (key ? (row[key] || "").trim() : "");

  for (const row of rows) {
    const name = g(row, cols.name);
    if (!name) continue;
    const rawType = cols.type ? (row[cols.type] || "").trim().toLowerCase() : "";
    const itype = VALID_TYPES.has(rawType) ? rawType : InstrumentType.OTHER;

    const qty = cols.quantity ? toFloat(row[cols.quantity]) : null;
    const curval = cols.current_value ? toFloat(row[cols.current_value]) : null;
    if (curval === null && qty === null) {
      result.warn(`Skipped ${name}: no quantity or current value.`, name);
      continue;
    }
    const invested = cols.invested_value ? toFloat(row[cols.invested_value]) : null;

    const txns: ParsedTxn[] = [];
    const start = cols.start_date ? parseDate(row[cols.start_date]) : null;
    if (start && invested) txns.push({ date: start, kind: "buy", amount: round2(-invested) });

    result.holdings.push(holding({
      name,
      instrument_type: itype,
      isin: g(row, cols.isin).toUpperCase() || null,
      symbol: g(row, cols.symbol).toUpperCase() || null,
      scheme_code: g(row, cols.scheme_code) || null,
      quantity: qty ?? 1.0,
      avg_cost: cols.avg_cost ? toFloat(row[cols.avg_cost]) : null,
      invested_value: invested,
      current_value: curval,
      amc: g(row, cols.amc) || null,
      plan: g(row, cols.plan).toLowerCase() || null,
      folio: g(row, cols.folio) || null,
      source: Source.GENERIC_CSV,
      account_name: accountName,
      account_kind: "manual",
      sector_hint: g(row, cols.sector) || null,
      market_cap_hint: g(row, cols.market_cap) || null,
      category_hint: g(row, cols.asset_class) || null,
      transactions: txns,
      raw: row,
    }));
  }
  result.info(`Parsed ${result.holdings.length} rows from generic CSV.`);
  return result;
}
